src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -448,83 +504,64 @@ protected internal void OnSolutionReloaded(SolutionInfo reloadedSolutionInfo)
-
-        private void OnProjectAdded(ProjectInfo projectInfo, bool silent)
-        {
-            using (_serializationLock.DisposableWait())
-            {
-                this.OnProjectAdded_NoLock(projectInfo, silent);
-            }
-        }
-
-        private void OnProjectAdded_NoLock(ProjectInfo projectInfo, bool silent)
         {
-            var projectId = projectInfo.Id;
-
-            CheckProjectIsNotInCurrentSolution(projectId);
-
-            var oldSolution = this.CurrentSolution;
-            var newSolution = this.SetCurrentSolution(oldSolution.AddProject(projectInfo));
-
-            if (!silent)
-            {
-                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectAdded, oldSolution, newSolution, projectId);
-            }
+            this.SetCurrentSolution(
+                oldSolution => CheckAndAddProject(oldSolution, projectInfo),
+                WorkspaceChangeKind.ProjectAdded, projectId: projectInfo.Id);
         }
 
         /// <summary>
         /// Call this method to respond to a project being reloaded in the host environment.
         /// </summary>
         protected internal virtual void OnProjectReloaded(ProjectInfo reloadedProjectInfo)
         {
-            using (_serializationLock.DisposableWait())
-            {
-                var projectId = reloadedProjectInfo.Id;
-
-                CheckProjectIsInCurrentSolution(projectId);
-
-                var oldSolution = this.CurrentSolution;
-                var newSolution = oldSolution.RemoveProject(projectId).AddProject(reloadedProjectInfo);
-
-                newSolution = this.AdjustReloadedProject(oldSolution.GetRequiredProject(projectId), newSolution.GetRequiredProject(projectId)).Solution;
-                newSolution = this.SetCurrentSolution(newSolution);
+            var projectId = reloadedProjectInfo.Id;
+            this.SetCurrentSolution(
+                oldSolution =>
+                {
+                    CheckProjectIsInSolution(oldSolution, projectId);
 
-                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectReloaded, oldSolution, newSolution, projectId);
-            }
+                    return this.AdjustReloadedProject(
+                        oldSolution.GetRequiredProject(projectId),
+                        oldSolution.RemoveProject(projectId).AddProject(reloadedProjectInfo).GetRequiredProject(projectId)).Solution;
+                }, WorkspaceChangeKind.ProjectReloaded, projectId);
         }
 
         /// <summary>
         /// Call this method to respond to a project being removed from the host environment.
         /// </summary>
         protected internal virtual void OnProjectRemoved(ProjectId projectId)
         {
+            // This currently doesn't use the SetCurrentSolution(transform) pattern as it changes mutable state (open
+            // docs), and as such needs to atomically change both that and the solution-snapshot.
```

Comment:
See earlier: can't you use your on after action to clear out the mutable state here?